File: apps/renderer/src/main/services/analytics.ts
import { join } from 'path'
import { readFileSync, writeFileSync, existsSync } from 'fs'
import { randomUUID } from 'crypto'
import { app } from 'electron'
import { paths } from './paths'
import { getConfig } from './config'

// ── Credentials ────────────────────────────────────────────────────────────
// The Measurement ID is public (it ships in every GA-instrumented client), so
// it's hardcoded. The Measurement Protocol API secret must NOT live in this
// public repo, so it's injected at build time from the GA_API_SECRET env via
// the main-process `define` in electron.vite.config.ts. Until the secret is
// provided the module stays inert (no network calls). The secret only grants
// event ingestion to this one property.
declare const __GA_API_SECRET__: string | undefined

const MEASUREMENT_ID = 'G-MX248BTV0T'
const API_SECRET = typeof __GA_API_SECRET__ === 'string' ? __GA_API_SECRET__ : ''
const ENDPOINT = `https://www.google-analytics.com/mp/collect?measurement_id=${MEASUREMENT_ID}&api_secret=${API_SECRET}`

// The Measurement ID is always present; analytics only "go live" once the
// build-injected API secret is set.
const configured = (): boolean => API_SECRET !== ''
// Opt-out: analytics runs unless the user explicitly disabled it.
const consented = (): boolean => getConfig().analyticsEnabled !== false

type EventParams = Record<string, string | number>

// A stable, anonymous per-install id (a random UUID — not tied to any account,
// device fingerprint, or PII). Persisted next to the other app data.
let clientId = ''
const sessionId = String(Date.now())

function clientIdFile(): string { return join(paths.userData, 'analytics.json') }

function loadClientId(): string {
  try {
    const data = JSON.parse(readFileSync(clientIdFile(), 'utf-8')) as { clientId?: string }
    if (data.clientId) return data.clientId
  } catch { /* fall through to generate */ }
  const id = randomUUID()
  try { writeFileSync(clientIdFile(), JSON.stringify({ clientId: id }), 'utf-8') } catch { /* best-effort */ }
  return id
}

/** Initialise analytics and record the app_open event. Safe to call always. */
export function initAnalytics(): void {
  if (!configured()) return
  clientId = loadClientId()
  trackEvent('app_open')
}

/**
 * Send a GA4 event. No-op unless analytics is both configured and consented.
 * Only pass non-identifying parameters — never usernames, UUIDs, tokens, or
 * file paths.
 */
export function trackEvent(name: string, params: EventParams = {}): void {
  if (!configured() || !consented()) return
  if (!clientId) clientId = loadClientId()

  const body = JSON.stringify({
    client_id: clientId,
    events: [{
      name,
      params: {
        // session_id + engagement_time_msec are what make GA4 count the event
        // as an active session (otherwise Realtime/engagement stay empty).
        session_id: sessionId,
        engagement_time_msec: 100,
        app_version: app.getVersion(),
        os: process.platform,
        ...params,
      },
    }],
  })

  // Fire-and-forget. Telemetry must never surface an error to the user, and we
  // deliberately don't log failures here (the logger forwards errors back into
  // analytics, which would recurse).
  void fetch(ENDPOINT, { method: 'POST', body }).catch(() => { /* ignore */ })
}

File: apps/tauri/src-tauri/src/analytics.rs
use crate::{config, paths};
use serde_json::{json, Value};
use std::collections::HashMap;
use std::fs;
use std::sync::OnceLock;
use std::time::{SystemTime, UNIX_EPOCH};
use uuid::Uuid;

const MEASUREMENT_ID: &str = "G-MX248BTV0T";

fn configured() -> bool {
    !api_secret().is_empty()
}

fn api_secret() -> &'static str {
    option_env!("GA_API_SECRET").unwrap_or("")
}

fn consented() -> bool {
    config::read()
        .get("analyticsEnabled")
        .and_then(Value::as_bool)
        != Some(false)
}

fn client_id_file() -> std::path::PathBuf {
    paths::data_dir().join("analytics.json")
}

fn session_id() -> &'static str {
    static SESSION_ID: OnceLock<String> = OnceLock::new();
    SESSION_ID.get_or_init(|| {
        SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|duration| duration.as_millis().to_string())
            .unwrap_or_else(|_| "0".to_string())
    })
}

fn load_client_id() -> String {
    if let Ok(text) = fs::read_to_string(client_id_file()) {
        if let Ok(value) = serde_json::from_str::<Value>(&text) {
            if let Some(id) = value.get("clientId").and_then(Value::as_str) {
                return id.to_string();
            }
        }
    }

    let id = Uuid::new_v4().to_string();
    let _ = fs::create_dir_all(paths::data_dir());
    let _ = fs::write(client_id_file(), json!({ "clientId": id }).to_string());
    id
}

fn sanitize_params(params: Option<Value>) -> HashMap<String, Value> {
    let mut safe = HashMap::new();
    let Some(Value::Object(map)) = params else {
        return safe;
    };

    for (key, value) in map {
        if key.len() > 40 || !key.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
            continue;
        }
        match value {
            Value::String(text) if text.len() <= 120 => {
                safe.insert(key, Value::String(text));
            }
            Value::Number(number) => {
                safe.insert(key, Value::Number(number));
            }
            _ => {}
        }
    }

    safe
}

pub fn init() {
    track_event("app_open", None);
}

pub fn track_event(name: &str, params: Option<Value>) {
    if !configured() || !consented() {
        return;
    }

    let allowed = matches!(
        name,
        "app_open" | "page_view" | "instance_launch" | "app_error"
    );
    if !allowed {
        return;
    }

    let mut event_params = sanitize_params(params);
    event_params.insert("session_id".into(), json!(session_id()));
    event_params.insert("engagement_time_msec".into(), json!(100));
    event_params.insert("app_version".into(), json!(env!("CARGO_PKG_VERSION")));
    event_params.insert("os".into(), json!(std::env::consts::OS));

    let endpoint = format!(
        "https://www.google-analytics.com/mp/collect?measurement_id={MEASUREMENT_ID}&api_secret={}",
        api_secret()
    );
    let body = json!({
        "client_id": load_client_id(),
        "events": [{
            "name": name,
            "params": event_params,
        }],
    });

    tauri::async_runtime::spawn(async move {
        let _ = reqwest::Client::new()
            .post(endpoint)
            .json(&body)
            .send()
            .await;
    });
}

#[tauri::command]
pub fn analytics_track(name: String, params: Option<Value>) {
    if name == "page_view" {
        track_event(&name, params);
    }
}
